be/src/storage/lake/pk_index_loader.cpp

```diff
@@ -0,0 +1,200 @@
+        }
+        (*seg_itr)->close();
+    }
+
+private:
+    Tablet* _tablet;
+    uint32_t _rowset_id;
+    uint32_t _seg_id;
+    const std::string _seg_name;
+    const Schema _schema;
+    int64_t _version;
+    const MetaFileBuilder* _builder;
+};
+
+PkIndexLoader::~PkIndexLoader() {
+    _contexts.clear();
+}
+
+Status PkIndexLoader::load(Tablet* tablet, const std::vector<RowsetPtr>& rowsets, const Schema& schema, int64_t version,
+                           const MetaFileBuilder* builder) {
+    uint64_t subtask_num = 0;
+    std::vector<std::shared_ptr<PkSegmentScanTask>> scan_tasks;
+    for (auto& rowset : rowsets) {
+        auto& rowset_meta = rowset->metadata();
+        for (uint32_t i = 0; i < rowset_meta.segments_size(); ++i) {
+            auto scan_task = std::make_shared<PkSegmentScanTask>(tablet, rowset->id(), i, rowset_meta.segments(i),
+                                                                 schema, version, builder);
+            bool ret = ExecEnv::GetInstance()->connector_scan_executor()->submit(
+                    workgroup::ScanTask(workgroup::WorkGroupManager::instance()->get_default_workgroup().get(),
+                                        [scan_task]() { scan_task->run(); }));
+            if (!ret) {
+                auto error_msg = fmt::format("Fail to submit pk segment scan subtask, tablet: {}", tablet->id());
+                LOG(ERROR) << error_msg;
+                return Status::IOError(error_msg);
+            }
+        }
+        subtask_num += rowset_meta.segments_size();
+    }
+
+    std::lock_guard<std::mutex> lg(_mutex);
+    auto context = std::make_shared<PkSegmentScanTaskContext>();
+    context->subtask_num = subtask_num;
+    _contexts.emplace(tablet->id(), context);
+    return Status::OK();
+}
+
+void PkIndexLoader::add_chunk(int64_t tablet_id, const std::shared_ptr<ChunkInfo>& chunk) {
+    std::lock_guard<std::mutex> lg(_mutex);
+    auto it = _contexts.find(tablet_id);
+    if (it == _contexts.end()) {
+        return;
+    } else {
+        auto context = it->second;
+        context->chunk_infos.push(chunk);
+        context->cv.notify_all();
+    }
+}
+
+StatusOr<std::shared_ptr<ChunkInfo>> PkIndexLoader::get_chunk(int64_t tablet_id) {
+    std::unique_lock<std::mutex> lk(_mutex);
+    auto it = _contexts.find(tablet_id);
+    if (it == _contexts.end()) {
+        auto error_msg = fmt::format("Invalid tablet id {}", tablet_id);
+        return Status::NotFound(error_msg);
+    }
+
+    auto context = it->second;
+    if (context->subtask_num == 0 && context->chunk_infos.empty()) {
+        _contexts.erase(tablet_id);
+        return Status::EndOfFile("done");
+    }
+
+    if (!context->status.is_ok_or_eof()) {
+        Status st = context->status;
+        _contexts.erase(tablet_id);
+        return st;
+    }
+
+    if (!context->chunk_infos.empty()) {
+        return context->get_chunk_info();
+    }
+
+    context->cv.wait(lk, [&] {
```

Comment:
use wait_for with timeout, favor for clean exit, otherwise if the cv missed the notify, it will never wake up.

Reply:
done